### DISCUSSION TITLE
INS AD/nonAD version

### DISCUSSION THREAD

**Discussion by mcacace** [idaholab/moose#26316] [Q&A Modules: Navier-Stokes] [answered]
Dear all, 
I'm trying to "upgrade" a NS INS example to its AD version (see whether this could speed up the solver). I (think I) put almost all things together, though I am stuck with a BC. The problem is that I need to impose a MatchValueBC along a specific boundary (whose values I take from an auxVariable - lagrange), and I'm not sure how to make it working while having the velocity as a lagrange vector variable. I tried to have a lagrange auxvariable (with 3 components) and make use of the BuildArrayVariable, but it is not compatible with lagrange_vec type. This said, before looking at a solution on my own, I though it wise to ask the forum...  
I am attaching my two input files (sorry if this violates the posting rules), for anyone who could help.
Thanks a lot 
[fracture_flow.zip](https://github.com/idaholab/moose/files/13657948/fracture_flow.zip)

**Comment by GiudGiud** [comment]
Hello

So you need a VectorMatchedValueBC?
Or do you need a way to create a vector variable, to then use it in a VectorMatchedValueBC?

Guillaume

**Comment by mcacace** [comment]
In principle a way to create a VectorVariable (same as per the velocity - Lagrange_vec) to use in a VectorMatchedValueBC

**Reply by GiudGiud** [reply, marked as answer]
Does the ParsedVectorAux let you do that ?

Or is it restricted to non nodal variables as well

**Reply by mcacace** [reply]
I can give it a try (guess I can move the auxvariables to nodes). Will report back tomorrow once in the office. 
Thanks again

**Reply by GiudGiud** [reply]
sounds good. Thanks for engaging on this

BuildArrayVariable is for array variables but maybe we need to add a BuildVectorVariable to make it clear what we're doing with vector variables

**Reply by mcacace** [reply]
@GiudGiud your hints did work indeed! Thanks a lot.

**Comment by mcacace** [comment]
Of course, the vector variable should be populated from existing lagrangian variables.
Thanks

**Comment by lindsayad** [comment]
> I'm trying to "upgrade" a NS INS example to its AD version (see whether this could speed up the solver)

If you're referring to the INS vs INSAD prefixed objects in the `navier_stokes` module you're unlikely to see better solver performance with INSAD unless you're running on displacing meshes. We worked pretty hard to get the Jacobians correct for the INS objects

**Reply by mcacace** [reply]
I see, indeed we are running on a displaced mesh. And, just tested it, did not provide any speed up wrt the INS formulation, so in the end I will stay with the non AD version. Thanks anyway for the support.

**Reply by lindsayad** [reply]
Are you using PJFNK or NEWTON?
